File: real_estate/enhancements/project.py
import frappe
from frappe import _
from real_estate.real_estate.doctype.property_trigger_type.property_trigger_type import get_valid_project_triggers
from erpnext.projects.doctype.project.project import Project
from frappe.utils import today

class PropertyProject(Project):
	def onload(self):
		self.set_onload('valid_project_triggers', get_valid_project_triggers())


@frappe.whitelist()
def create_trigger(project, project_trigger, trigger_date=None):
	if not (project and project_trigger):
		return

	if not trigger_date:
		trigger_date = today()

	if frappe.db.exists('Property Trigger Table', {'parent': project, 'parenttype': 'Project',
		'trigger_type': project_trigger, 'trigger_date': trigger_date}):
		frappe.throw(_('Trigger on this date already exist'))

	doc = frappe.new_doc('Property Trigger Table')
	doc.parent = project
	doc.parenttype = 'Project'
	doc.parentfield = 'property_triggers'
	doc.trigger_type = project_trigger
	doc.trigger_date = trigger_date
	doc.insert()

File: real_estate/public/js/controllers/project.js
frappe.provide("real_estate");
frappe.provide('erpnext.projects');

real_estate.ProjectControllerExtended = class ProjectController extends erpnext.projects.ProjectController {
	onload() {
		super.onload();
	}

	refresh() {
		this.setup_buttons();
	}

	setup_buttons() {
//		super.setup_buttons();
		var me = this;

		if (!me.frm.is_new() && me.frm.doc.is_real_estate_project) {
			if (me.frm.doc.__onload && me.frm.doc.__onload.valid_project_triggers) {
				$.each(me.frm.doc.__onload.valid_project_triggers || [], function (i, project_trigger) {
					me.frm.add_custom_button(__(project_trigger), () => {
						me.create_trigger(me.frm, project_trigger);
					}, __('Create Trigger'));
				});
			}
		}
	}

	create_trigger(frm, project_trigger) {
		var dialog = new frappe.ui.Dialog({
			title: __("Triggering Invoice"),
			fields: [
				{
					fieldname: "trigger_type", fieldtype: "Link", options: "Property Trigger Type",
					label: "Trigger Type", default: project_trigger, read_only: 1,
				},
				{
					fieldtype: "Column Break"
				},
				{
					fieldname: "trigger_date", fieldtype: "Date", options: "",
					label: "Date", default: '', reqd: 1,
				},
			],
			primary_action_label: 'Create Trigger',
			primary_action(values) {

				frappe.call({
					method: "real_estate.enhancements.project.create_trigger",
					args: {
						project: frm.doc.name,
						project_trigger: values.trigger_type,
						trigger_date: values.trigger_date
					},
					callback: function(r) {
						if (r.message && !r.exc) {

						}
					}
				});
				dialog.hide();
				frm.reload_doc();
			}
		});
		dialog.show();
	}
 };

extend_cscript(cur_frm.cscript, new real_estate.ProjectControllerExtended({frm: cur_frm}));
